extension-workspace-reference.md

```diff
@@ -0,0 +1,121 @@
+# workspace/xreference extension to LSP
+
+The `workspace/xreference` extension to the Language Server Protocol (LSP) enables a language server to export all of the references to dependencies that code inside of the workspace makes.
+
+Use case: clients of a language server can invoke `workspace/xreference` in order to find external references to dependencies. This information can then be stored inside of a database, which allows the caller to create a 'global mapping' of symbols in dependencies to the workspace they are used in (e.g. to see "how do other people use this symbol?").
+
+### Initialization
+
+`ServerCapabilities` may contain a new field to indicate server-side support for this extension:
+
+```typescript
+interface ServerCapabilities {
+  /* ... all fields from the base ServerCapabilities ... */
+
+  /**
+   * The server provides workspace reference exporting support.
+   */
+  xworkspaceReferenceProvider?: boolean;
+}
+```
+
+#### Workspace Reference Request
+
+The workspace reference request is sent from the client to the server to export project-wide references to dependencies. That is, the response strictly returns references in the project to symbols defined in dependencies.
```

Comment:
@keegancsmith your question is somewhat ambiguous to me. The Go language server needs to understand what/where a dependency is, yes, e.g. that someone is importing a dependency that is vendored. We do this today. It doesn't need intimate knowledge about vendor tools, unless those break the standard assumptions of the Go vendor thing / require special casing.

Unlike the old `workspace/reference` implementation, this new one effectively says "any reference to a dependency". So that's any `pkg.Symbol` in Go -- regardless of whether or not `pkg` is vendored or not. The only thing the Go LS must include is whether or not that package is vendored (so that a user query could include/exclude vendor packages from their results). No more funny business with "defined in the workspace" etc is included with the new implementation.